Add ASG lifecycle management Lambda function

joshbeard opened the pull request:
## Description

This introduces an Auto Scaling Group instance termination lifecycle
hook using Lambda and related resources. The Lambda function is a Python
script that is triggered when the persistent runner instance in the ASG
is terminated. The function receives the instance ID of the "parent"
runner and queries for spawned instances that it launched to terminate.
Additionally, it will check for other "orphaned" instances that have a
`gitlab-runner-parent-id` tag that doesn't match an existing instance. This
resolves the issue where spawned instances could be orphaned when their
parent runner is terminated.

This feature is disabled by default.

The user data script is updated to provide the 'parent' instance ID as a
tag named 'gitlab-runner-parent-id' on spawned instances.

A new sub-module is provided called "terminate-workers". It is optional
to use this feature, and the input variable
`asg_terminate_lifecycle_hook_create` can be toggled `true` or `false`
for this behavior.

This partially addresses concerns discussed in issue #214

## Migrations required

NO

## Verification

I've tested existing standard configuration with this addition _disabled_ and _enabled_ with expected results.

* On a default setup without enabling this feature, there are 2 changes - the user data is updated to provide the `gitlab-runner` and `gitlab-runner-parent` tags to spawned worker instances.

    ```
    Plan: 0 to add, 2 to change, 0 to destroy.
    ````

* When enabled with its default configuration, there are 10 resource additions and the two changes mentioned previously.

    ```
    Plan: 10 to add, 2 to change, 0 to destroy.
    ```

## Documentation

We use [pre-commit](https://pre-commit.com/) to update the Terraform inputs and outputs in the documentation via [terraform-docs](https://github.com/terraform-docs/terraform-docs). Ensure you have installed those components.

Documentation was updated, tflint and pre-commit hooks ran.

joshbeard commented:
Thanks for the review, @kayman-mk! I'll work on fixes and adjustments from your feedback. The Python requirement comes from the use of an external module (https://registry.terraform.io/modules/terraform-aws-modules/lambda/) for the Lambda function and is mentioned as a prerequisite when using this feature in the documentation updates in this pull request. However, I think we can get rid of this and handle this simple use case in this module. I'll do some refactoring for this.

On the issue of it not actually doing anything - I'm curious if you can check the CloudWatch event rule to see if it triggered an invocation and if it failed? There's also a CloudWatch log stream for the function itself. On the worker instance, do you see the `gitlab-runner-parent` tag with a value of the parent instance ID?

joshbeard commented:
@npalm rebased

joshbeard commented:
> @npalm Thank you so much for the thorough review and consideration! I will resolve the issues you pointed out, rebase, and push within the next couple of days.

Couple of days = couple of weeks ;) I've been away for the past couple of weeks.

kayman-mk commented on main.tf:
```diff
@@ -503,3 +503,24 @@ resource "aws_iam_role_policy_attachment" "eip" {
+### Lambda function for ASG instance termination lifecycle hook
+################################################################################
+module "terminate_workers_lifecycle_function" {
+  source = "./modules/terminate-workers"
```
question: Should we rename this to `terminate-runner-instances`? I checked the the Gitlab documentation and they have a Runner (the agent or coordinator running 24/7 and requesting new jobs) and Runner Instances (the machines doing the work). I am also not very precise with the wording in my PRs, but I think it would be better to do so. This should be done for all other occurrences as well.

joshbeard replied:
I replaced 'workers' with 'instances' in a290efa4934c4a43c3f14143be6f21d5ba1ef689

kayman-mk commented on main.tf:
```diff
@@ -503,3 +503,24 @@ resource "aws_iam_role_policy_attachment" "eip" {
+module "terminate_workers_lifecycle_function" {
+  source = "./modules/terminate-workers"
+
+  count = var.asg_terminate_lifecycle_hook_create ? 1 : 0
```
question: Should we really make this configurable? As far as I understand this PR (and also from my daily work with this module) it fixes a bug. Thus it should always be enabled to make sure that there are no dangling instances. @npalm

npalm replied:
Yes I would suggest to make it configurable, but with default enabled. We don't know who is using the module but some orgs are very restrictive for lambda, for example enforcing using lambda's in a custom VPC. By making the the feature optiona (opt-out) we do not block anyone to use the module

joshbeard replied:
Pushed an update to enable this by default

kayman-mk commented on modules/terminate-workers/variables.tf:
```diff
@@ -0,0 +1,71 @@
+  type        = string
+}
+
+variable "create" {
```
question: I am not 100% sure how to use this variable. What is the use case not to create the resources? This results in a module which does not the work it was written for, right? Wouldn't I get the same result in not creating the whole module as every resource has a `count = var.create ? 1 :0`?

joshbeard replied:
You're right - this is useless in the submodule. I've removed it in a290efa4934c4a43c3f14143be6f21d5ba1ef689

kayman-mk commented on modules/terminate-workers/iam.tf:
```diff
@@ -0,0 +1,103 @@
+resource "aws_iam_role" "lambda" {
+  count = var.create ? 1 : 0
+
+  name                  = var.name_iam_objects == "" ? var.name : var.name_iam_objects
```
question: According to the variable description `name_iam_objects` it sounds to me that the name of this and all other IAM objects should be something like `name = "${var.name_iam_objects]-suffix"`

joshbeard replied:
Fixed the IAM resource names to be consistent and use the module's "environment" variable in a290efa4934c4a43c3f14143be6f21d5ba1ef689

kayman-mk commented on modules/terminate-workers/lambda/lambda_function.py:
```diff
@@ -0,0 +1,120 @@
+    ec2_instances = client.describe_instances(Filters=[
+            {
+                "Name": "instance-state-name",
+                "Values": ['running'],
```
question: Is this sufficient? What if a new instance is being created? We should kill it too.

joshbeard replied:
I added "pending" to this list in a290efa4934c4a43c3f14143be6f21d5ba1ef689

kayman-mk commented on modules/terminate-workers/lambda/lambda_function.py:
```diff
@@ -0,0 +1,120 @@
+                        _terminate_list.append(instance['InstanceId'])
+                    else:
+                        try:
+                            # Handle other workers without a parent that are still running.
```
praise: Good job to handle these instances too!

kayman-mk commented on modules/terminate-workers/iam.tf:
```diff
@@ -0,0 +1,103 @@
+      "ec2:TerminateInstances"
+    ]
+    resources = ["*"]
+    condition {
```
praise: Showing the concept of least privileges. Nice!

kayman-mk commented on modules/terminate-workers/versions.tf:
```diff
@@ -0,0 +1,7 @@
+  required_version = ">= 0.13"
+  required_providers {
+    aws   = ">= 2.46"
+    local = ">= 1.4"
```
question: Is this provider still in use?

joshbeard replied:
Nope, removed in a290efa4934c4a43c3f14143be6f21d5ba1ef689

joshbeard commented on modules/terminate-workers/README.md:
```diff
@@ -0,0 +1,184 @@
+module "runner" {
+  source  = "npalm/gitlab-runner/aws"
+
+  aws_region                    = "us-west-2"
```
Resolved: https://github.com/npalm/terraform-aws-gitlab-runner/pull/392/commits/28a7488c63089fdd888fba61fd08036886a3e973#diff-93dca4429ba0587d39551de9d4aa0822dfd70511e62075c5949c6c5d5e07cf68R52

joshbeard commented on modules/terminate-workers/variables.tf:
```diff
@@ -0,0 +1,71 @@
+# ----------------------------------------------------------------------------
+# Terminate Workers - Input Variables
+# ----------------------------------------------------------------------------
+variable "name" {
```
I've added this in 28a7488c63089fdd888fba61fd08036886a3e973

joshbeard commented on modules/terminate-workers/cloudwatch.tf:
```diff
@@ -0,0 +1,39 @@
+resource "aws_cloudwatch_event_rule" "terminate_instances" {
+  count = var.create ? 1 : 0
+
+  name        = var.name
```
Added in 28a7488c63089fdd888fba61fd08036886a3e973

joshbeard commented on modules/terminate-workers/main.tf:
```diff
@@ -0,0 +1,68 @@
+  source_code_hash = data.archive_file.terminate_runner_workers_lambda.output_base64sha256
+  function_name    = var.name
+  handler          = "lambda_function.handler"
+  memory_size      = 128
```
I added a variable for this in https://github.com/npalm/terraform-aws-gitlab-runner/pull/392/commits/a290efa4934c4a43c3f14143be6f21d5ba1ef689

joshbeard commented on modules/terminate-workers/variables.tf:
```diff
@@ -0,0 +1,71 @@
+  default     = {}
+}
+
+variable "local_existing_package" {
```
Removed in a290efa4934c4a43c3f14143be6f21d5ba1ef689

joshbeard commented on modules/terminate-workers/variables.tf:
```diff
@@ -0,0 +1,71 @@
+  default     = null
+}
+
+variable "source_path" {
```
Removed in a290efa4934c4a43c3f14143be6f21d5ba1ef689

joshbeard commented:
@npalm @kayman-mk I've finally gotten back around to this and have made updates based on your most recent feedback in commit a290efa4934c4a43c3f14143be6f21d5ba1ef689.

I noticed the `README` is updated with some unrelated changes from the pre-commit hook. Not sure why. I left it as a separate commit for now to make it easier to adjust if you want me to remove the unrelated changes (which still seem to be accurate?) That's visible specifically in ab8a2b79af35a0dff4741b0c2d2ef2cb5afc5cc3

I'll rebase this after follow-up reviews and potential tweaks.

npalm commented:
Great work, added to my backlog.

npalm commented:
Will check in the weekend

kayman-mk commented:
@joshbeard  @npalm I did a quick check in my environment.
- Deploying the module: Success
- Killing the runner manually: The executor was killed by the Lambda function. Success
- Didn't find any suspicious log entries in the Lambda log.
- Executors have a new tag: `gitlab-runner-parent-id`

What I noticed:
- Lambda timeout is 3s. This might be too short as the duration of the call was 2s.
- I have to enable the new function via `asg_terminate_lifecycle_hook_create`. This doesn't feel handy. I think it is a base functionality which should always be enabled to be sure not to waste money.
- Log messages are JSON formatted and have a structured field "Level". Really appreciated. Makes log analysis easy.

joshbeard commented:
> @kayman-mk I also will do a check. Would suggest to make the lambda memory as well the time out configurable with a reasonable default

@npalm I've made the lambda memory size and the timeout configurable: https://github.com/npalm/terraform-aws-gitlab-runner/pull/392/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288e

joshbeard commented:
Rebased against "develop"

GertVil commented:
Really looking forward to having this merged, because this was one of the "issues" I documented before we could look into using this module at work. Do you have any idea when this can be merged @npalm? (Thank you for the awesome work!)

npalm commented:
> Really looking forward to having this merged, because this was one of the "issues" I documented before we could look into using this module at work. Do you have any idea when this can be merged @npalm? (Thank you for the awesome work!)

@GertVil I will try to check the PR tonight. Can you provide a short direction how to test the lambda?

GertVil commented:
> > Really looking forward to having this merged, because this was one of the "issues" I documented before we could look into using this module at work. Do you have any idea when this can be merged @npalm? (Thank you for the awesome work!)
> 
> @GertVil I will try to check the PR tonight. Can you provide a short direction how to test the lambda?

I'm afraid that @joshbeard will probably be in a better position to provide you with that information since he created all this so he probably has a scenario that he has used before.

npalm commented on variables.tf:
```diff
@@ -734,3 +734,33 @@ variable "subnet_ids_gitlab_runner" {
+
+variable "asg_terminate_lifecycle_lambda_timeout" {
+  description = "Amount of time the terminate-instances Lambda Function has to run in seconds."
+  default     = 10
```
Would it not be better to set the time out a bit higher? 30 for example? The lambda was running for me 2 to 3 seconds, but only 1 instance to kill.

joshbeard replied:
I increased the default to 30

npalm commented on modules/terminate-instances/cloudwatch.tf:
```diff
@@ -0,0 +1,33 @@
+
+resource "aws_cloudwatch_event_target" "terminate_instances" {
+  rule      = aws_cloudwatch_event_rule.terminate_instances.name
+  target_id = "TriggerTerminateLambda"
```
According the docs this should be a unique identifier, if not set a random will be created. So I suggest either to remove it or add the environment name to make it unique.
```suggestion
  target_id = "${var.environment}-TriggerTerminateLambda"
```

npalm reviewed (changes requested):
@joshbeard great job! works like a charm! Only one small change request

joshbeard commented:
> @joshbeard great job! works like a charm! Only one small change request, see inline. Can you also run a rebase please.

@npalm Adjustments made and rebased against 'develop'

npalm commented:
Sorry, missed the update. Will merge the PR tomorrow. If you have time for a rebase, please. Othewrise I will fix it locally!

joshbeard commented:
Rebased again. The README changed a bit since my last rebase, as I'm sure you're aware. I added the resources to the "resources" section, but that may get overwritten with the current pre-commit config.

npalm commented:
@joshbeard thx, checking. Just a qucik tip for a next time. Fore reviewing it helps if you are not squashing, only rebase. Now I have to go over the full PR again. No worries we get this merged!

npalm reviewed (approved):
@joshbeard All good, thanks!!!

joshbeard commented:
> @joshbeard thx, checking. Just a qucik tip for a next time. Fore reviewing it helps if you are not squashing, only rebase. Now I have to go over the full PR again. No worries we get this merged! 

I'm sorry about that. Thanks for the review.

joshbeard commented:
> @joshbeard All good, thanks!!!

Great! Thanks for your work on this, the review, feedback, and including this in the project!
